examples/pipeline_wavernn/main.py

```diff
@@ -0,0 +1,420 @@
+        val_dataset,
+        batch_size=args.batch_size,
+        collate_fn=collate_fn,
+        **loader_validation_params,
+    )
+
+    model = _WaveRNN(
+        upsample_scales=args.upsample_scales,
+        n_bits=args.n_bits,
+        sample_rate=args.sample_rate,
+        hop_length=args.hop_length,
+        n_res_block=args.n_res_block,
+        n_rnn=args.n_rnn,
+        n_fc=args.n_fc,
+        kernel_size=args.kernel_size,
+        n_freq=args.n_freq,
+        n_hidden=args.n_hidden,
+        n_output=args.n_output,
+        mode=args.mode,
+    )
+
+    if args.jit:
+        model = torch.jit.script(model)
+
+    model = torch.nn.DataParallel(model)
+    model = model.to(devices[0], non_blocking=True)
+
+    n = count_parameters(model)
+    logging.info(f"Number of parameters: {n}")
+
+    # Optimizer
+    optimizer_params = {
+        "lr": args.learning_rate,
+    }
+
+    optimizer = Adam(model.parameters(), **optimizer_params)
+
+    criterion = nn.CrossEntropyLoss() if args.mode == "waveform" else MoLLoss
+
+    best_loss = 10.0
+
+    load_checkpoint = args.checkpoint and os.path.isfile(args.checkpoint)
+
+    if load_checkpoint:
+        logging.info(f"Checkpoint: loading '{args.checkpoint}'")
+        checkpoint = torch.load(args.checkpoint)
+
+        args.start_epoch = checkpoint["epoch"]
+        best_loss = checkpoint["best_loss"]
+
+        model.load_state_dict(checkpoint["state_dict"])
+        optimizer.load_state_dict(checkpoint["optimizer"])
+
+        logging.info(
+            f"Checkpoint: loaded '{args.checkpoint}' at epoch {checkpoint['epoch']}"
+        )
+    else:
+        logging.info("Checkpoint: not found")
+
+        save_checkpoint(
+            {
+                "epoch": args.start_epoch,
+                "state_dict": model.state_dict(),
+                "best_loss": best_loss,
+                "optimizer": optimizer.state_dict(),
+            },
+            False,
+            args.checkpoint,
+        )
+
+    for epoch in range(args.start_epoch, args.epochs):
+
+        train_one_epoch(
+            model, args.mode, criterion, optimizer, train_loader, devices[0], epoch,
+        )
+
+        if not (epoch + 1) % args.print_freq or epoch == args.epochs - 1:
+
+            sum_loss = validate(
+                model, args.mode, criterion, val_loader, devices[0], epoch,
+            )
+
+            is_best = sum_loss < best_loss
+            best_loss = min(sum_loss, best_loss)
+            save_checkpoint(
+                {
+                    "epoch": epoch + 1,
+                    "state_dict": model.state_dict(),
+                    "best_loss": best_loss,
+                    "optimizer": optimizer.state_dict(),
+                },
+                is_best,
+                args.checkpoint,
+            )
+
+    logging.info(f"End time: {datetime.now()}")
+
+
+if __name__ == "__main__":
+
+    logging.basicConfig(level=logging.INFO)
+    args = parse_args()
+    main(args)
```

Comment:
you could add a final function after `main` that takes the save the final weights as you have done [here](https://github.com/jimchen90/TestRep/blob/master/wavernn/load_weights.ipynb).

Reply:
let's not do this in the example pipeline to keep it minimal